crates/iceberg/src/writer/partitioning/clustered.rs

```diff
@@ -0,0 +1,251 @@
+// // Licensed to the Apache Software Foundation (ASF) under one
+// // or more contributor license agreements.  See the NOTICE file
+// // distributed with this work for additional information
+// // regarding copyright ownership.  The ASF licenses this file
+// // to you under the Apache License, Version 2.0 (the
+// // "License"); you may not use this file except in compliance
+// // with the License.  You may obtain a copy of the License at
+// //
+// //   http://www.apache.org/licenses/LICENSE-2.0
+// //
+// // Unless required by applicable law or agreed to in writing,
+// // software distributed under the License is distributed on an
+// // "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// // KIND, either express or implied.  See the License for the
+// // specific language governing permissions and limitations
+// // under the License.
+//
+// //! This module provides the `ClusteredWriter` implementation.
+//
+// use std::collections::HashSet;
+//
+// use arrow_array::RecordBatch;
+// use async_trait::async_trait;
+//
+// use crate::spec::{PartitionKey, Struct};
+// use crate::writer::{DefaultInput, DefaultOutput, IcebergWriter, IcebergWriterBuilder};
+// use crate::writer::file_writer::location_generator::{FileNameGenerator, LocationGenerator};
+// use crate::{Error, ErrorKind, Result};
+// use crate::writer::partitioning::PartitioningWriter;
+//
+// /// A writer that writes data to a single partition at a time.
+// ///
+// /// When data for a new partition arrives, it closes the current writer
+// /// and creates a new one for the new partition.
+// ///
+// /// Once a partition has been written to and closed, any further attempts
+// /// to write to that partition will result in an error.
+// pub struct ClusteredWriter<B: IcebergWriterBuilder, I: Default + Send = DefaultInput, O: Default + Send = DefaultOutput>
```

Comment:
Please ignore this for now, I think it's better to keep this draft/round of changes focused on the interfaces changes with existing writer